designs/2022-eslint-to-esm/README.md

```diff
@@ -0,0 +1,212 @@
+            "import": "./lib/unsupported-api.js",
+            "require": "./dist/unsupported-api.cjs",
+        }
+    }
+```
+
+Node.js ESM has a limitation: ESM cannot be required. Approach 2 provides a good compatibility for existing projects that rely on ESLint without requiring users to adopt ESM as well. In previous repos, dual-mode was mostly used (except for '@eslint/create-config').
+
+However, dual-mode also has a few disadvantages:
+1. dependencies. Some ESLint's dependencies have been (or will be) esm-only, which means we won't be able to upgrade them if we continue to support CJS. For new dependencies, we cannot use them if they are esm-only.
+2. Slower installation & larger volume usage, even though most users only use ESM, they still need to download CJS modules.
+3. Higher maintenance costs. You need to add a build process, test and publish ESM/CJS code.
+4. The CLIs do not support dual-mode.
+
+In the (very) long run, I would suggest that we eventually go with esm-only, but to give users more time to upgrade, as an intermediate state, we could use dual-mode in ESLint v9:
+
+* eslint v9:
+    - ESLint CLI: esm-only
+    - Node.js API: dual mode
+* eslint 1x (at the point when the majority of the ecosystem is ready to support it, likely to be v10/v11/...):
+    - ESLint CLI: esm-only
+    - Node.js API: esm-only
+
+## Toolings esm supports
+
+* mocha ✅
+* proxyquire ❌
+* nyc ❌
+* eslint-plugin-node ❗️(Not fully supported, but we could use a fork: `eslint-plugin-n`)
+
+## Implementations
+1. ESLint v9
+* package.json + `"type": "module"`
+* package.json exports
+* package.json engines >= Node.js v12
+* remove `"use strict"`
+* convert `"./foo"` to `"./foo/index.js"`
+* convert `"./foo"` to `"./foo.js"`
+* convert `require/module.exports` => `import/export`
+* esm toolings(`nyc` -> `c8`, `proxyquire` -> `esmock`)
+* cjs supports(build `dist/api.cjs` and `dist/use-at-your-own-risk.cjs` using rollup)
+* update docs to use esm
+
+2. ESLint 1x
+* remove package.json exports.cjs (use esm-only)
+* remove build process
+* upgrade blocked deps.
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+Yes. When ESLint v9 is released, it is recommended to announce that users should upgrade to ESM, CJS may be no longer supported in future major versions.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think 
+    about any opposing viewpoints that reviewers might bring up. 
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+1. Compatibility issues, see below.
+
+## Backwards Compatibility Analysis
```

Comment:
Would be good to call out the impact to third-party plugin developers. Can plugins write rules/configs in either CJS/ESM? Can plugins be entirely ESM? What about rule tests? Currently they use: `const RuleTester = require('eslint').RuleTester;`

Reply:
This will be a non-issue going forward. With the new config system it doesn’t matter how you write your plugins because you’re in charge of loading them.

Reply:
@simlu fwiw, the better way to effectively tree-shake is to granularly import only what you need - it does a better job than any treeshaking possibly can :-)

Reply:
@ljharb LOL, better even, custom rewrite everything so it's the minimal code you need 😜 /sarcasmend

Sorry I should have been more concise. I was talking about ncc. So not traditional tree shaking and what you import really doesn't matter (but yes we do that already)